garden/src/components/form/garden/edit/EditGardenForm.tsx

```diff
@@ -0,0 +1,74 @@
+import { Garden, GardenPatchRequest } from "@/api/types";
+import { useForm, useFormContext } from "react-hook-form";
+import { Form } from "@/components/ui/form";
+import { formSchema, GardenPatchFormData } from "./schemas";
+import { getDirtyValues } from "@/lib/utils";
+import { zodResolver } from "@hookform/resolvers/zod";
+import React from "react";
+import { usePatchGarden } from "@/api";
+import FormFields from "./FormFields";
+import { UnsavedChangesDialog } from "../../UnsavedChangesDialog";
+import { useBlocker } from "react-router-dom";
+import LoadingSpinner from "@/components/LoadingSpinner";
+
+export const EditGardenForm = ({ garden }: { garden: Garden }) => {
+  const { mutateAsync: patchGarden } = usePatchGarden();
+
+  const form = useForm<GardenPatchFormData>({
+    resolver: zodResolver(formSchema),
+    mode: "onTouched",
+    defaultValues: React.useMemo(
+      () => ({
+        title: garden.title || "",
+        description: garden.description || "",
+        year: garden.year || "",
+        authors: garden.authors || [],
+        version: garden.version || "",
+        contributors: garden.contributors || [],
+        tags: garden.tags || [],
+      }),
+      [garden],
+    ),
+  });
+
+  const blocker = useBlocker(
+    () => !form?.formState.isSubmitting && Object.keys(form.formState.touchedFields).length > 0,
+  );
+
+  const onSubmit = React.useCallback(
+    async (values: GardenPatchFormData) => {
+      const gardenPatchRequest: GardenPatchRequest = getDirtyValues(
+        values,
+        form.formState.dirtyFields,
+      ) as GardenPatchRequest;
+
+      await patchGarden({
+        doi: garden.doi,
+        garden: gardenPatchRequest,
+      });
+    },
+    [garden.doi, patchGarden],
+  );
+
+  return (
+    <form onSubmit={form.handleSubmit(onSubmit)}>
+      <Form {...form}>
+        <FormFields />
```

Comment:
At some point I could use a walkthrough to understand our form libraries a little better. It seems like react-hook-form is used at the "lower" level and hookform is used more at the edges for validation against the zod schema. Is that right at a high level or am I way off base?

Reply:
Yeah, the way I understand it, hookform supplies different validators (like Zod). I think react-hook-form has its own built-in validator, but Zod is pretty clean/intuitive so I've always used that instead

React hook form supplies all of the nice form controls you'd want, removing the hassle of having a bunch of useState/useEffect for fields/errors. It keeps track of registered fields, errors, other form state, and when the form is submitted, uses the validator to make sure the fields match the expected schema. Really nice for things 'field Arrays' ([check out the "Associated Materials" section of entrypoint editing](https://github.com/Garden-AI/garden-frontend/blob/staging/garden/src/components/form/entrypoint/edit/AssociatedMaterialsSection.tsx)), which are fields that are arrays that can grow/shrink in size